### ISSUE TITLE
False positive: error CS1643: Not all code paths return a value in lambda expression of type 'Func<int>'

### ISSUE THREAD

**Issue by bernd5** [dotnet/roslyn#52612]
**Version Used**: https://github.com/dotnet/roslyn/commit/4e4c8237665adb00a0d0c5f710b9f5ce1a2086d6

**Steps to Reproduce**:
[Compile:](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEBMAjALABQZAbhAgAQQDGMNAvDZnAO424CsAPAEtMMAHwAKMQEpmIgN5kAkFA4CY9ABY0xAIigB7MHBjqhAc22T5pBQvrQ4NKrQD6NDuriYahGgKiPqGj12EEUbXAB2bwBuRQBfMjjJaKA==)

```CSharp
using System;

var act = new Func<int>(()=>{
	switch ("something"){
		case var _ when 1 is var one:
			return one;
	}
});
```

**Expected Behavior**:
No error message.

**Actual Behavior**:
error CS1643: Not all code paths return a value in lambda expression of type 'Func<int>'
That a var pattern always succeeds seems to be ignored here.

**Comment by jaredpar** [comment]
Thanks for taking the time to file this issue but this behavior is "By Design". Once a `when` clause comes into play in a `switch` expression the compiler assumes that can return `false`. Yes we could optimize this specific case to know it won't return `false` but that was not in the scope of the feature.

**Comment by bernd5** [comment]
I checked it - I don't think it is true. Because:

```
using System;

var act = new Func<int>(()=>{
	switch ("something"){
		case var _ when true:
			return 1;
	}
});
```
compiles just fine.
But `1 is var one` has not a constant value for roslyn.
I assume because of potential side effects - okay for me.
